src/express-lane-auction/ExpressLaneAuction.sol

```diff
@@ -0,0 +1,477 @@
+    function initiateWithdrawal() external {
+        uint64 curRnd = currentRound();
+        uint256 amount = _balanceOf[msg.sender].balance;
+        _balanceOf[msg.sender].initiateWithdrawal(curRnd);
+        // CHRIS: TODO: we have + 2 here, that's leaking an implementation detail
+        emit WithdrawalInitiated(msg.sender, amount, curRnd + 2);
+    }
+
+    /// @inheritdoc IExpressLaneAuction
+    function finalizeWithdrawal() external {
+        uint256 amountReduced = _balanceOf[msg.sender].finalizeWithdrawal(currentRound());
+        biddingToken.safeTransfer(msg.sender, amountReduced);
+        // CHRIS: TODO: consider adding the following assertion - it's an invariant
+        // CHRIS: TODO: Invariant: assert(withdrawableBalance(msg.sender) == 0);
+        emit WithdrawalFinalized(msg.sender, amountReduced);
+    }
+
+    /// @dev Update local state to resolve an auction
+    /// @param isMultiBid Where the auction should be resolved from multiple bids
+    /// @param firstPriceBid The winning bid
+    /// @param firstPriceBidder The winning bidder
+    /// @param priceToPay The price that needs to be paid by the winner
+    /// @param biddingInRound The round bidding is taking place in. This is not the round the bidding is taking place for, which is biddingInRound + 1
+    function resolveAuction(
+        bool isMultiBid,
+        Bid calldata firstPriceBid,
+        address firstPriceBidder,
+        uint256 priceToPay,
+        uint64 biddingInRound
+    ) internal {
+        // store that a round has been resolved
+        uint64 biddingForRound = biddingInRound + 1;
+        latestResolvedRounds.setResolvedRound(biddingForRound, firstPriceBid.expressLaneController);
+
+        // first price bidder pays the beneficiary
+        _balanceOf[firstPriceBidder].reduce(priceToPay, biddingInRound);
+        biddingToken.safeTransfer(beneficiary, priceToPay);
+
+        // emit events so that the offchain sequencer knows a new express lane controller has been selected
+        (uint64 roundStart, uint64 roundEnd) = roundTimingInfo.roundTimestamps(biddingForRound);
+        emit SetExpressLaneController(
+            biddingForRound,
+            address(0),
+            firstPriceBid.expressLaneController,
+            roundStart,
+            roundEnd
+        );
+        emit AuctionResolved(
+            isMultiBid,
+            biddingForRound,
+            firstPriceBidder,
+            firstPriceBid.expressLaneController,
+            firstPriceBid.amount,
+            priceToPay,
+            roundStart,
+            roundEnd
+        );
+    }
+
+    /// @inheritdoc IExpressLaneAuction
+    function getBidBytes(
+        uint64 _round,
+        uint256 _amount,
+        address _expressLaneController
+    ) public view returns (bytes memory) {
+        // CHRIS: TODO: test the length of this is 112
+        return
+            abi.encodePacked(block.chainid, address(this), _round, _amount, _expressLaneController);
```

Comment:
shoud use EIP712 typed hash with domain separator instead of custom schema

Reply:
is it worth doing this since people wont be signing this in user facing wallets?

Reply:
will add as a TODO, might switch to this later